src/background/services/openPayments.ts

```diff
@@ -424,31 +424,46 @@ export class OpenPaymentsService {
     if (calculatedHash !== hash) throw new Error('Invalid interaction hash')
   }
 
-  private async confirmPayment(url: string): Promise<InteractionParams> {
+  private async closeTab(currentTab: number, tabToClose: number) {
+    await this.browser.tabs.update(currentTab, { active: true })
+    await this.browser.tabs.remove(tabToClose)
+  }
+
+  private async getInteractionInfo(url: string): Promise<InteractionParams> {
     const currentTab = await getCurrentActiveTab(this.browser)
 
     return await new Promise((res) => {
-      if (url) {
-        this.browser.tabs.create({ url }).then((tab) => {
-          if (tab.id) {
-            this.browser.tabs.onUpdated.addListener((tabId, changeInfo) => {
-              try {
-                const tabUrl = new URL(changeInfo.url || '')
-                const interactRef = tabUrl.searchParams.get('interact_ref')
-                const hash = tabUrl.searchParams.get('hash')
-
-                if (tabId === tab.id && interactRef && hash) {
-                  this.browser.tabs.update(currentTab.id, { active: true })
-                  this.browser.tabs.remove(tab.id)
-                  res({ interactRef, hash })
-                }
-              } catch (e) {
-                /* do nothing */
+      this.browser.tabs.create({ url }).then((tab) => {
+        if (tab.id) {
+          const getInteractionInfo: Parameters<
+            Tabs.onUpdatedEvent['addListener']
+          >[0] = async (tabId, changeInfo) => {
+            if (tabId !== tab.id) return
+            try {
+              const tabUrl = new URL(changeInfo.url || '')
+              const interactRef = tabUrl.searchParams.get('interact_ref')
+              const hash = tabUrl.searchParams.get('hash')
+              const result = tabUrl.searchParams.get('result')
+
+              if (
+                (interactRef && hash) ||
+                result === 'grant_rejected' ||
+                result === 'grant_invalid'
+              ) {
+                await this.closeTab(currentTab.id!, tabId)
+                this.browser.tabs.onUpdated.removeListener(getInteractionInfo)
               }
-            })
+
+              if (interactRef && hash) {
+                res({ interactRef, hash })
```

Comment:
Should we reject in other case? And handle it in popup?

Reply:
Let's discuss this in today's call and we can have another PR that takes care of this. Some time ago I was thinking about using a real web page - instead of using `localhost:3344` -  as the redirect URL.